src/token.rs

```diff
@@ -0,0 +1,221 @@
+use std::fmt;
+
+use diesel;
+use diesel::prelude::*;
+use diesel::pg::PgConnection;
+use conduit::{Request, Response};
+use time::Timespec;
+use conduit_router::RequestParams;
+use rustc_serialize::json;
+
+use db::RequestTransaction;
+use user::{RequestUser, AuthenticationSource};
+use util::{RequestUtils, CargoError, CargoResult, ChainError, human, read_fill};
+use schema::api_tokens;
+
+/// The model representing a row in the `api_tokens` database table.
+#[derive(Clone, Debug, PartialEq, Eq, Identifiable, Queryable)]
+pub struct ApiToken {
+    pub id: i32,
+    pub user_id: i32,
+    pub token: String,
+    pub name: String,
+    pub created_at: Timespec,
+    pub last_used_at: Option<Timespec>,
+}
+
+/// The serialization format for the `ApiToken` model.
+#[derive(RustcDecodable, RustcEncodable)]
+pub struct EncodableApiToken {
+    pub id: i32,
+    pub name: String,
+    pub token: Option<String>,
+    pub created_at: String,
+    pub last_used_at: Option<String>,
+}
+
+impl ApiToken {
+    /// Generates a new named API token for a user
+    pub fn insert(conn: &PgConnection, user_id: i32, name: &str) -> CargoResult<ApiToken> {
+        #[table_name = "api_tokens"]
+        #[derive(Insertable, AsChangeset)]
+        struct NewApiToken<'a> {
+            user_id: i32,
+            name: &'a str,
+        }
+
+        diesel::insert(&NewApiToken {
+            user_id: user_id,
+            name: name,
+        }).into(api_tokens::table)
+            .get_result::<ApiToken>(conn)
+            .map_err(From::from)
+    }
+
+    /// Deletes the provided API token if it belongs to the provided user
+    pub fn delete(conn: &PgConnection, user_id: i32, id: i32) -> CargoResult<()> {
+        diesel::delete(api_tokens::table.find(id).filter(
+            api_tokens::user_id.eq(user_id),
+        )).execute(conn)?;
+        Ok(())
+    }
+
+    pub fn find_for_user(conn: &PgConnection, user_id: i32) -> CargoResult<Vec<ApiToken>> {
+        api_tokens::table
+            .filter(api_tokens::user_id.eq(user_id))
+            .order(api_tokens::created_at.desc())
+            .load::<ApiToken>(conn)
+            .map_err(From::from)
+    }
+
+    pub fn count_for_user(conn: &PgConnection, user_id: i32) -> CargoResult<u64> {
+        api_tokens::table
+            .filter(api_tokens::user_id.eq(user_id))
+            .count()
+            .get_result::<i64>(conn)
+            .map(|count| count as u64)
+            .map_err(From::from)
+    }
+
+    /// Converts this `ApiToken` model into an `EncodableApiToken` for JSON
+    /// serialization.
+    pub fn encodable(self) -> EncodableApiToken {
```

Comment:
Is this method separate from `encodable_with_token` for security reasons? What do you think about separating them into separate types to make future mistakes less likely, and further clarify intent?